### PR TITLE
Initial commit of nmcli: NetworkManager module. Currently supports:

### PR DESCRIPTION
**Pull request by alcamie101**
Create, modify, remove of - team, team-slave, bond, bond-slave, ethernet
TODO:
vlan, bridge, wireless related connections.
more docs here: https://github.com/alcamie101/ansible-nmcli
Why did I write an Ansible module for NetworkManager?
The main reason I started writing this module was because I was building or should I say rebuilding an OpenStack undercloud the N'th time and I was sick of having to write Jinja templates for configuring the interfaces and then write all of the supporting playbook guff to manage deployment and consequent network management shell cmds for one type of interface and then rewrite it again for another type of interface.

This makes things simple, quick and extremely flexible. E.g. You can build multiple bond interfaces with assorted NIC configurations and deploy them within a couple of minutes. If you're not happy with bonding, you can, in a few seconds, change the configuration to team and team-slaves and redeploy that in a couple of minutes just by changing the 'type'. At the moment, the playbook process here would be:

Create bond
create bond-slave
deploy 'playbook-add.yml'
decide you want teams not bonds
Remove above connection 'playbook-del.yml'
in 'playbook-add.yml' change bond to team and bond-slave to team-slave
deploy 'playbook-add.yml' and voila, you now have teams rather than bonds.

### CODE REVIEW

File: network/nmcli.py [hunk 1]
```diff
@@ -0,0 +1,1089 @@
...
+module: nmcli
+author: Chris Long
+short_description: Manage Networking
+requirements: [ nmcli, dbus ]
+description:
+    - Manage the network devices. Create, modify, and manage, ethernet, teams, bonds, vlans etc.
+options:
+    state:
+        required: True
+        default: "present"
```
Comment (bcoca) [inline network/nmcli.py]: when required is true, there is no default

File: network/nmcli.py [hunk 2]
```diff
@@ -0,0 +1,1089 @@
...
+    description:
+        - Whether the device should exist or not, taking action if the state is different from what is stated.
+    enabled:
+        required: False
+        default: "yes"
+        choices: [ "yes", "no" ]
+        description:
+            - Whether the service should start on boot. B(At least one of state and enabled are required.)
+            - Whether the connection profile can be automatically activated ( default: yes)
+    action:
```
Comment (bcoca) [inline network/nmcli.py]: we shy away from action/command type of interfaces as they are imperative and not declarative

add/modify seem to overlap with state present
delte with state absent

up/down with enabled yes/no or it's own flag up: yes/no

show looks like an attempt to make this a facts plugin

if you shift enabled to 'autoconnect true/false' then we can use it as i described above

File: network/nmcli.py [hunk 3]
```diff
@@ -0,0 +1,1089 @@
...
+                 70: "IP Config",
+                 80: "IP Check",
+                 90: "Secondaries",
+                 100: "Activated",
+                 110: "Deactivating",
+                 120: "Failed"
+            }
+
+    def __new__(cls, *args, **kwargs):
+        return load_platform_subclass(Nmcli, args, kwargs)
```
Comment (bcoca) [inline network/nmcli.py]: where is load_platform_subclass defined?

File: network/nmcli.py [hunk 4]
```diff
@@ -0,0 +1,1089 @@
...
+            flags=dict(required=False, default=None, type='str'),
+            ingress=dict(required=False, default=None, type='str'),
+            egress=dict(required=False, default=None, type='str'),
+        ),
+        supports_check_mode=True
+    )
+
+    nmcli=Nmcli(module)
+
+    if nmcli.syslogging:
```
Comment (bcoca) [inline network/nmcli.py]: ansible already outputs to syslog, no need to do yourself

File: network/nmcli.py [hunk 5]
```diff
@@ -0,0 +1,1089 @@
...
+        default: None
+        choices: [ add, modify, show, up, down ]
+        description:
+            - Set to 'add' if you want to add a connection.
+            - Set to 'modify' if you want to modify a connection. Modify one or more properties in the connection profile.
+            - Set to 'delete' if you want to delete a connection. Delete a configured connection. The connection to be deleted is identified by its name 'cfname'.
+            - Set to 'show' if you want to show a connection. Will show all devices unless 'cfname' is set.
+            - Set to 'up' if you want to bring a connection up. Requires 'cfname' to be set.
+            - Set to 'down' if you want to bring a connection down. Requires 'cfname' to be set.
+    cname:
```
Comment (bcoca) [inline network/nmcli.py]: description contradicts required/default

cname might be confusing with dns aliases, conn_name or something more explicit might be better

Comment (alcamie101) [inline network/nmcli.py]: cname is the name used in NetworkManagers own docs. I can change to conn_name if that's less confusing.

File: network/nmcli.py (hunk 1 above)
Comment (alcamie101) [inline network/nmcli.py]: will fix that. Thanks

File: network/nmcli.py (hunk 4 above)
Comment (alcamie101) [inline network/nmcli.py]: Ahh ok. I pulled this from the user module. Thanks for the insight.

File: network/nmcli.py [hunk 6]
```diff
@@ -0,0 +1,1089 @@
...
+            - This is the type of device or network connection that you wish to create for a bond, team or bridge. (NetworkManager default: balance-rr)
+    master:
+        required: False
+        default: None
+        description:
+            - master <master (ifname, or connection UUID or cname) of bridge, team, bond master connection profile.
+    ip4:
+        required: False
+        default: None
+        description: The IPv4 address to this interface using this format ie: "192.168.1.24/24"
```
Comment (bcoca) [inline network/nmcli.py]: descriptions should always be lists ( - prefix)

File: network/nmcli.py (hunk 3 above)
Comment (alcamie101) [inline network/nmcli.py]: Old reference, thanks for pointing it out.

File: network/nmcli.py (hunk 5 above)
Comment (bcoca) [inline network/nmcli.py]: you can always keep as an alias, but a less confusing default name is a better interface imho

File: network/nmcli.py [hunk 7]
```diff
@@ -0,0 +1,1089 @@
...
+description:
+    - Manage the network devices. Create, modify, and manage, ethernet, teams, bonds, vlans etc.
+options:
+    state:
+        required: True
+        default: "present"
+        choices: [ present, absent ]
+    description:
+        - Whether the device should exist or not, taking action if the state is different from what is stated.
+    enabled:
```
Comment (bcoca) [inline network/nmcli.py]: saying that the service is enabled or not is confusing and sounds like it belongs in the service module, after looking at code this sets the connection to autoconnect or not.

File: network/nmcli.py (hunk 2 above)
Comment (alcamie101) [inline network/nmcli.py]: have removed as I had added functions to defunct the need for actions

File: network/nmcli.py (hunk 6 above)
Comment (alcamie101) [inline network/nmcli.py]: correcting.

**Comment by alcamie101** [comment]
bcoca. I've corrected all of the action points and tested.

**Comment by alcamie101** [comment]
+1

**Pull request merged**